File: attribouter/src/main/java/me/jfenn/attribouter/wedges/TranslatorWedge.java
package me.jfenn.attribouter.wedges;

import android.content.Context;
import android.content.res.XmlResourceParser;
import androidx.annotation.Nullable;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import org.xmlpull.v1.XmlPullParserException;

import java.io.IOException;

import me.jfenn.attribouter.R;
import me.jfenn.attribouter.data.github.GitHubData;
import me.jfenn.attribouter.data.github.UserData;
import me.jfenn.attribouter.interfaces.Mergeable;
import me.jfenn.attribouter.utils.ResourceUtils;
import me.jfenn.attribouter.utils.UrlClickListener;

public class TranslatorWedge extends Wedge<TranslatorWedge.ViewHolder> implements Mergeable<TranslatorWedge> {

    @Nullable
    public String login;
    @Nullable
    public String name;
    @Nullable
    public String avatarUrl;
    @Nullable
    public String blog;
    @Nullable
    public String email;
    @Nullable
    public String locales;

    public TranslatorWedge(XmlResourceParser parser) throws IOException, XmlPullParserException {
        this(parser.getAttributeValue(null, "login"),
                parser.getAttributeValue(null, "name"),
                parser.getAttributeValue(null, "avatar"),
                parser.getAttributeValue(null, "locales"),
                parser.getAttributeValue(null, "blog"),
                parser.getAttributeValue(null, "email"));
    }

    protected TranslatorWedge(@Nullable String login, @Nullable String name, @Nullable String avatarUrl, @Nullable String locales, @Nullable String blog, @Nullable String email) {
        super(R.layout.item_attribouter_translator);
        this.login = login;
        this.name = name;
        this.avatarUrl = avatarUrl;
        this.locales = locales;
        this.blog = blog;
        this.email = email;

        if (login != null && !hasEverything())
            addRequest(new UserData(login));
    }

    @Override
    public void onInit(GitHubData data) {
        if (data instanceof UserData) {
            UserData user = (UserData) data;
            merge(new TranslatorWedge(
                    user.login,
                    user.name,
                    user.avatar_url,
                    null,
                    user.blog,
                    user.email
            ));
        }
    }

    @Nullable
    public String getName() {
        return name != null ? name : login;
    }

    @Override
    public TranslatorWedge merge(TranslatorWedge contributor) {
        if ((name == null || !name.startsWith("^")) && contributor.name != null)
            name = contributor.name;
        if ((avatarUrl == null || !avatarUrl.startsWith("^")) && contributor.avatarUrl != null)
            avatarUrl = contributor.avatarUrl;
        if ((blog == null || !blog.startsWith("^")) && contributor.blog != null && !contributor.blog.isEmpty())
            blog = contributor.blog;
        if ((email == null || !email.startsWith("^")) && contributor.email != null && !contributor.email.isEmpty())
            email = contributor.email;
        if ((locales == null || !locales.startsWith("^")) && contributor.locales != null)
            locales = contributor.locales;

        return this;
    }

    @Override
    public boolean hasAll() {
        return false;
    }

    @Override
    public boolean isHidden() {
        return false;
    }

    public boolean hasEverything() {
        return name != null && name.startsWith("^") && blog != null && blog.startsWith("^");
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof TranslatorWedge) {
            TranslatorWedge translator = (TranslatorWedge) obj;
            return (login != null && translator.login != null && login.toLowerCase().equals(translator.login.toLowerCase())) || super.equals(obj);
        } else return super.equals(obj);
    }

    @Override
    public ViewHolder getViewHolder(View v) {
        return new ViewHolder(v);
    }

    @Override
    public void bind(Context context, ViewHolder viewHolder) {
        ResourceUtils.setImage(context, avatarUrl, viewHolder.imageView);
        viewHolder.nameView.setText(ResourceUtils.getString(context, getName()));

        String blog = ResourceUtils.getString(context, this.blog);
        if (blog != null) {
            viewHolder.itemView.setOnClickListener(new UrlClickListener(blog));
        } else if (login != null) {
            viewHolder.itemView.setOnClickListener(new UrlClickListener("https://github.com/" + login));
        } else viewHolder.itemView.setOnClickListener(null);
    }

    protected static class ViewHolder extends Wedge.ViewHolder {

        protected ImageView imageView;
        protected TextView nameView;

        protected ViewHolder(View v) {
            super(v);
            imageView = v.findViewById(R.id.image);
            nameView = v.findViewById(R.id.name);
        }
    }
}

File: attribouter/src/main/java/me/jfenn/attribouter/wedges/TranslatorWedge.kt
package me.jfenn.attribouter.wedges

import android.content.Context
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import me.jfenn.attribouter.R
import me.jfenn.attribouter.utils.*
import me.jfenn.gitrest.model.ProviderString
import me.jfenn.gitrest.model.User
import java.util.*

open class TranslatorWedge(
        login: String? = null,
        name: String? = null,
        avatarUrl: String? = null,
        locales: String? = null,
        blog: String? = null,
        email: String? = null
) : Wedge<TranslatorWedge.ViewHolder>(R.layout.attribouter_item_translator) {

    var login: String? by attr("login", login)
    var name: String? by attr("name", name)
    var avatar: String? by attr("avatar", avatarUrl)
    var websiteUrl: String? by attr("websiteUrl", blog)
    var locales: String? by attr("locales", locales)
    var email: String? by attr("email", email)

    // whether the translator is the first of a locale in the sorted list (i.e. should display its header)
    var isFirst: Boolean = false

    override fun onCreate() {
        login?.let {
            if (!hasEverything()) lifecycle?.launch {
                withContext(Dispatchers.IO) {
                    lifecycle?.client?.getUser(it)
                }?.let { user -> onTranslator(user) }
            }
        }
    }

    open fun onTranslator(data: User) {
        name = data.name
        avatar = data.avatarUrl
        websiteUrl = data.websiteUrl
        email = data.email

        notifyItemChanged()
    }

    open fun getDisplayName(): String? {
        return name ?: login
    }

    private fun hasEverything(): Boolean {
        return !name.isResourceMutable() && !websiteUrl.isResourceMutable()
    }

    override fun equals(other: Any?): Boolean {
        return (other as? TranslatorWedge)?.let {
            login.equalsProvider(other.login)
        } ?: super.equals(other)
    }

    fun clone() : TranslatorWedge {
        return TranslatorWedge(
                login,
                name,
                avatar,
                locales,
                websiteUrl,
                email
        ).also {
            it.addChildren(getChildren())
        }
    }

    override fun getViewHolder(v: View): ViewHolder {
        return ViewHolder(v)
    }

    override fun bind(context: Context, viewHolder: ViewHolder) {
        viewHolder.localeView?.apply {
            text = locales?.split(",")?.getOrNull(0)?.let {
                Locale(it).getDisplayName(Locale(it))
            } ?: ""
            visibility = if (isFirst) View.VISIBLE else View.GONE
        }

        viewHolder.imageView?.apply {
            context.loadDrawable(avatar, R.drawable.attribouter_image_avatar) {
                setImageDrawable(it)
            }
        }

        viewHolder.nameView?.apply {
            text = ResourceUtils.getString(context, getDisplayName())
        }

        viewHolder.itemView.apply {
            websiteUrl?.let {
                setOnClickListener(UrlClickListener(it))
            } ?: login?.let {
                setOnClickListener(UrlClickListener(ProviderString(it).inferUrl()))
            } ?: run {
                setOnClickListener(null)
            }
        }
    }

    open class ViewHolder(v: View) : Wedge.ViewHolder(v) {
        var localeView: TextView? = v.findViewById(R.id.locale)
        var imageView: ImageView? = v.findViewById(R.id.image)
        var nameView: TextView? = v.findViewById(R.id.name)
    }
}
